nifi-api/src/main/java/org/apache/nifi/authorization/AccessPolicy.java

```diff
@@ -0,0 +1,291 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.authorization;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Defines a policy for a set of entities to perform a set of actions on a given resource.
+ */
+public class AccessPolicy {
+
+    private final String identifier;
+
+    private final Resource resource;
+
+    private final Set<String> entities;
+
+    private final Set<RequestAction> actions;
+
+    private AccessPolicy(final AccessPolicyBuilder builder) {
+        this.identifier = builder.identifier;
+        this.resource = builder.resource;
+
+        Set<String> entities = new HashSet<>();
+        if (builder.entities != null) {
+            entities.addAll(builder.entities);
+        }
+        this.entities = Collections.unmodifiableSet(entities);
+
+        Set<RequestAction> actions = new HashSet<>();
+        if (builder.actions != null) {
+            actions.addAll(builder.actions);
+        }
+        this.actions = Collections.unmodifiableSet(actions);
+
+        if (this.identifier == null || this.identifier.trim().isEmpty()) {
+            throw new IllegalArgumentException("Identifier can not be null or empty");
+        }
+
+        if (this.resource == null) {
+            throw new IllegalArgumentException("Resource can not be null");
+        }
+
+        if (this.entities == null || this.entities.isEmpty()) {
+            throw new IllegalArgumentException("Entities can not be null or empty");
+        }
+
+        if (this.actions == null || this.actions.isEmpty()) {
+            throw new IllegalArgumentException("Actions can not be null or empty");
+        }
+    }
+
+    /**
+     * @return the identifier for this policy
+     */
+    public String getIdentifier() {
+        return identifier;
+    }
+
+    /**
+     * @return the resource for this policy
+     */
+    public Resource getResource() {
+        return resource;
+    }
+
+    /**
+     * @return an unmodifiable set of entity ids for this policy
+     */
+    public Set<String> getEntities() {
+        return entities;
+    }
+
+    /**
+     * @return an unmodifiable set of actions for this policy
+     */
+    public Set<RequestAction> getActions() {
+        return actions;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+
+        final AccessPolicy other = (AccessPolicy) obj;
+        return Objects.equals(this.identifier, other.identifier);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(this.identifier);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("identifier[%s], resource[%s], entityId[%s], action[%s]",
```

Comment:
I think there is a trailing ", " in the String.format. This applies the toString() in User, Group, and AccessPolicy.